api/ccapi/include/model.h

```diff
@@ -192,6 +195,20 @@ class Model {
     const std::vector<std::string> &end_layers, ReferenceLayersType type,
     const std::vector<std::string> &type_properties = {}) = 0;
 
+  /**
+   * @brief Visit each layer inside model
+   * @param fn function to be called for each layer
+   * - call param
+   *   @param layer layer the call back is visiting
+   *   @param rc run layer context clubbed with layer, null if not initialized
+   * yet
+   *
+   */
+  virtual void
+  forEachLayer(std::function<void(Layer & /**< layer */,
```

Comment:
You should note a few things with opening this API:
- layer_context.h must be opened with this
- add a `void *user_data` which will passed to forEachLayer, and will be included in the function signature.
- if this is being opened, it would be good to also open `registerCallback` with 3 modes (before forwarding, before backwarding, after backwarding) which will also receive RunLayerContext

Regarding this particular function, if not initialized, I recommend to not allowed to be run for the layers as finalize of the layers is itself not run as each layer will have to check for this.

Reply:
> layer_context.h must be opened with this

yes, layer_context.h is also used for the custom devel so already opened. We will need to move it to ml::train {}.

> add a void *user_data which will be passed to forEachLayer

std::function already can capture states so we won't need this. This will be needed for the capi front.

> it would be good to also open registerCallback with 3 modes 

Yeah, we will need to have some entry point for the model api. Let's discuss this when fixating model c++ api.

> Regarding this particular function, if not initialized, I recommend not allow to be run for the layers as finalize of the layers is itself not run as each layer will have to check for this. As the name/properties of the layer, ordering of execution, count of layers, and model itself change before and after compilation, it would be tough for the user to understand what is happening.

I think this is fair point. I'll not allow this and update documentation as per suggestion.

Reply:
> > add a void *user_data which will be passed to forEachLayer

> std::function already can capture states so we won't need this. This will be needed for the capi front.

If there will be a `void *user_data` for the c API anyway, I recommend to add here itself.

Reply:
IMO, adding void * user_data is likely to encourage bad practice to the c++ api user. There is a functor, which is type safe and broader way to alter void *.

Let's have in-depth discussion later though, I'll add `void *` for now :)

Reply:
cool.